File: src/image_resizer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import io
import os
import tempfile
from typing import Any, Dict

from PIL import Image


class ImageResizer:
    """
    画像リサイズ処理を行うクラス
    SNS別のサイズ制限に応じて画像を最適化する
    """

    # SNS別の画像制限設定
    SNS_LIMITS = {
        'bluesky': {
            'max_file_size': 1024 * 1024,  # 1MB
            'max_width': 2000,
            'max_height': 2000,
            'quality': 85
        },
        'x': {
            'max_file_size': 5 * 1024 * 1024,  # 5MB
            'max_width': 4096,
            'max_height': 4096,
            'quality': 85
        },
        'mastodon': {
            'max_file_size': 10 * 1024 * 1024,  # 10MB（一般的な設定）
            'max_width': 1920,
            'max_height': 1920,
            'quality': 85
        },
        'misskey': {
            'max_file_size': 10 * 1024 * 1024,  # 10MB（一般的な設定）
            'max_width': 2048,
            'max_height': 2048,
            'quality': 85
        }
    }

    def __init__(self, debug: bool = False):
        self.debug = debug

    def _debug_print(self, message: str):
        """デバッグメッセージを出力"""
        if self.debug:
            print(f"[DEBUG] ImageResizer: {message}")

    def resize_image_data(self, image_data: bytes, sns_type: str = 'bluesky') -> bytes:
        """
        画像データをSNS制限に合わせてリサイズする
        
        Args:
            image_data: 元画像のバイトデータ
            sns_type: SNSの種類（bluesky, x, mastodon, misskey）
            
        Returns:
            bytes: リサイズ後の画像データ
        """
        try:
            # SNS設定を取得
            limits = self.SNS_LIMITS.get(sns_type, self.SNS_LIMITS['bluesky'])
            max_size = limits['max_file_size']
            max_width = limits['max_width']
            max_height = limits['max_height']
            quality = limits['quality']

            self._debug_print(f"開始: {sns_type}, 元サイズ: {len(image_data)} bytes")

            # 元画像がサイズ制限内の場合はそのまま返す
            if len(image_data) <= max_size:
                image = Image.open(io.BytesIO(image_data))
                if image.width <= max_width and image.height <= max_height:
                    self._debug_print("リサイズ不要")
                    return image_data

            # 画像を開く
            image = Image.open(io.BytesIO(image_data))
            original_format = image.format

            # RGBAモードの場合はRGBに変換（JPEGサポートのため）
            if image.mode in ('RGBA', 'LA'):
                background = Image.new('RGB', image.size, (255, 255, 255))
                if image.mode == 'RGBA':
                    background.paste(image, mask=image.split()[-1])
                else:
                    background.paste(image)
                image = background
            elif image.mode != 'RGB':
                image = image.convert('RGB')

            self._debug_print(f"元画像: {image.width}x{image.height}, 形式: {original_format}")

            # アスペクト比を保持してリサイズ
            if image.width > max_width or image.height > max_height:
                image.thumbnail((max_width, max_height), Image.Resampling.LANCZOS)
                self._debug_print(f"サイズ調整後: {image.width}x{image.height}")

            # 品質を調整してファイルサイズを制限内に収める
            for current_quality in range(quality, 30, -5):  # 品質を段階的に下げる
                output = io.BytesIO()
                image.save(output, format='JPEG', quality=current_quality, optimize=True)
                result_data = output.getvalue()

                self._debug_print(f"品質{current_quality}: {len(result_data)} bytes")

                if len(result_data) <= max_size:
                    self._debug_print(f"完了: {len(result_data)} bytes (品質: {current_quality})")
                    return result_data

            # 最低品質でも制限を超える場合は、さらにサイズを縮小
            scale_factor = 0.9
            while len(result_data) > max_size and scale_factor > 0.5:
                new_width = int(image.width * scale_factor)
                new_height = int(image.height * scale_factor)
                resized = image.resize((new_width, new_height), Image.Resampling.LANCZOS)

                output = io.BytesIO()
                resized.save(output, format='JPEG', quality=30, optimize=True)
                result_data = output.getvalue()

                self._debug_print(f"追加縮小 {scale_factor:.1f}: {new_width}x{new_height}, {len(result_data)} bytes")

                if len(result_data) <= max_size:
                    break

                scale_factor -= 0.1

            self._debug_print(f"最終: {len(result_data)} bytes")
            return result_data

        except Exception as e:
            self._debug_print(f"リサイズエラー: {e}")
            # エラーの場合は元画像を返す
            return image_data

    def resize_image_file(self, file_path: str, sns_type: str = 'bluesky') -> str:
        """
        画像ファイルをSNS制限に合わせてリサイズし、一時ファイルとして保存
        
        Args:
            file_path: 元画像ファイルのパス
            sns_type: SNSの種類
            
        Returns:
            str: リサイズ後の一時ファイルパス
        """
        try:
            with open(file_path, 'rb') as f:
                original_data = f.read()

            resized_data = self.resize_image_data(original_data, sns_type)

            # 一時ファイルとして保存
            with tempfile.NamedTemporaryFile(delete=False, suffix='.jpg') as temp_file:
                temp_file.write(resized_data)
                temp_path = temp_file.name

            self._debug_print(f"一時ファイル作成: {temp_path}")
            return temp_path

        except Exception as e:
            self._debug_print(f"ファイルリサイズエラー: {e}")
            return file_path  # エラーの場合は元ファイルを返す

    def get_sns_limits(self, sns_type: str) -> Dict[str, Any]:
        """
        指定したSNSの制限情報を取得
        
        Args:
            sns_type: SNSの種類
            
        Returns:
            Dict[str, Any]: 制限情報
        """
        return self.SNS_LIMITS.get(sns_type, self.SNS_LIMITS['bluesky'])

    @staticmethod
    def cleanup_temp_file(file_path: str):
        """
        一時ファイルを削除
        
        Args:
            file_path: 削除する一時ファイルのパス
        """
        try:
            if file_path and os.path.exists(file_path) and file_path.startswith(tempfile.gettempdir()):
                os.unlink(file_path)
        except Exception:
            pass  # 削除エラーは無視


def create_image_resizer(debug: bool = False) -> ImageResizer:
    """
    ImageResizerインスタンスを作成
    
    Args:
        debug: デバッグモードの有効/無効
        
    Returns:
        ImageResizer: ImageResizerインスタンス
    """
    return ImageResizer(debug=debug)

File: src/image_resizer.rs
use image::codecs::jpeg::JpegEncoder;
use image::{DynamicImage, imageops};
use std::io::Cursor;

pub struct ImageResizer {
    debug: bool,
}

pub struct SnsLimits {
    pub max_file_size: usize,
    pub max_width: u32,
    pub max_height: u32,
    pub default_quality: u8,
}

impl ImageResizer {
    pub fn new(debug: bool) -> Self {
        Self { debug }
    }

    fn debug_print(&self, message: &str) {
        if self.debug {
            println!("[DEBUG] ImageResizer: {}", message);
        }
    }

    pub fn get_sns_limits(&self, sns_type: &str) -> SnsLimits {
        match sns_type {
            "bluesky" => SnsLimits {
                max_file_size: 1024 * 1024, // 1MB
                max_width: 2000,
                max_height: 2000,
                default_quality: 85,
            },
            "x" => SnsLimits {
                max_file_size: 5 * 1024 * 1024, // 5MB
                max_width: 4096,
                max_height: 4096,
                default_quality: 85,
            },
            "mastodon" => SnsLimits {
                max_file_size: 10 * 1024 * 1024, // 10MB
                max_width: 1920,
                max_height: 1920,
                default_quality: 85,
            },
            "misskey" => SnsLimits {
                max_file_size: 10 * 1024 * 1024, // 10MB
                max_width: 2048,
                max_height: 2048,
                default_quality: 85,
            },
            _ => SnsLimits {
                max_file_size: 1024 * 1024, // デフォルトはBluesky基準
                max_width: 2000,
                max_height: 2000,
                default_quality: 85,
            },
        }
    }

    pub fn resize_image_data(&self, image_data: &[u8], sns_type: &str) -> anyhow::Result<Vec<u8>> {
        let limits = self.get_sns_limits(sns_type);
        self.debug_print(&format!(
            "開始: {}, 元サイズ: {} bytes",
            sns_type,
            image_data.len()
        ));

        // 入力フォーマットを判定する(JPEG/PNG以外はJPEGへ再エンコードする方針のため)
        let format = image::guess_format(image_data).ok();

        // デコードする。失敗した場合(対応外形式: AVIF等)は元データを返さずエラーにする。
        // 元バイトをそのまま送るとSNS側で「media type unrecognized」等で弾かれ、
        // 何が起きたか分かりにくくなるため、ここで明確に失敗させて呼び出し側で
        // 画像なし投稿へフォールバックさせる。
        let img = image::load_from_memory(image_data).map_err(|e| {
            self.debug_print(&format!("画像デコード失敗: {}", e));
            anyhow::anyhow!("画像のデコードに失敗しました(対応外の形式の可能性): {}", e)
        })?;

        let width = img.width();
        let height = img.height();

        // JPEG/PNGで上限内ならそのまま返す。各SNSのupload_mimeはpng/jpegはそのまま、
        // それ以外はimage/jpegとして送るため、この2形式のみ無変換が安全。
        let is_jpeg_or_png = matches!(
            format,
            Some(image::ImageFormat::Jpeg) | Some(image::ImageFormat::Png)
        );
        if is_jpeg_or_png
            && image_data.len() <= limits.max_file_size
            && width <= limits.max_width
            && height <= limits.max_height
        {
            self.debug_print("変換不要(JPEG/PNGかつ上限内)");
            return Ok(image_data.to_vec());
        }

        // ここから先は常にJPEGとして出力する(WebP等の非JPEG/PNG、または上限超過時)。
        self.debug_print(&format!(
            "JPEGへ再エンコード: {}x{}, 形式: {:?}, color: {:?}",
            width,
            height,
            format,
            img.color()
        ));

        // 2. RGBA画像等の場合は白背景に重ねてRGBにする
        let rgb_img = match img {
            DynamicImage::ImageRgba8(rgba_img) => {
                let mut base = image::ImageBuffer::from_pixel(
                    rgba_img.width(),
                    rgba_img.height(),
                    image::Rgba([255, 255, 255, 255]),
                );
                imageops::overlay(&mut base, &rgba_img, 0, 0);
                DynamicImage::ImageRgba8(base).to_rgb8()
            }
            DynamicImage::ImageLumaA8(luma_a_img) => {
                let rgba_img = DynamicImage::ImageLumaA8(luma_a_img).to_rgba8();
                let mut base = image::ImageBuffer::from_pixel(
                    rgba_img.width(),
                    rgba_img.height(),
                    image::Rgba([255, 255, 255, 255]),
                );
                imageops::overlay(&mut base, &rgba_img, 0, 0);
                DynamicImage::ImageRgba8(base).to_rgb8()
            }
            other => other.to_rgb8(),
        };
        let mut rgb_img = DynamicImage::ImageRgb8(rgb_img);

        // 3. アスペクト比を維持して最大縦横に収まるよう縮小する
        if rgb_img.width() > limits.max_width || rgb_img.height() > limits.max_height {
            rgb_img = rgb_img.thumbnail(limits.max_width, limits.max_height);
            self.debug_print(&format!(
                "サイズ調整後: {}x{}",
                rgb_img.width(),
                rgb_img.height()
            ));
        }

        // 4. 品質を段階的に下げて圧縮を試みる
        let mut result_data = Vec::new();
        let mut current_quality = limits.default_quality;

        while current_quality >= 30 {
            let mut buf = Vec::new();
            let mut cursor = Cursor::new(&mut buf);
            let encoder = JpegEncoder::new_with_quality(&mut cursor, current_quality);
            if let Err(e) = rgb_img.write_with_encoder(encoder) {
                self.debug_print(&format!("JPEGエンコードエラー: {}", e));
                return Ok(image_data.to_vec());
            }
            result_data = buf;
            self.debug_print(&format!(
                "品質{}: {} bytes",
                current_quality,
                result_data.len()
            ));

            if result_data.len() <= limits.max_file_size {
                self.debug_print(&format!(
                    "完了: {} bytes (品質: {})",
                    result_data.len(),
                    current_quality
                ));
                return Ok(result_data);
            }

            if current_quality <= 30 {
                break;
            }
            current_quality = current_quality.saturating_sub(5);
        }

        // 5. 最低品質(30)でも超える場合は、さらにスケールを縮小
        let mut scale_factor = 0.9f32;
        while result_data.len() > limits.max_file_size && scale_factor > 0.5 {
            let new_width = (rgb_img.width() as f32 * scale_factor) as u32;
            let new_height = (rgb_img.height() as f32 * scale_factor) as u32;

            let resized = rgb_img.resize(new_width, new_height, imageops::FilterType::CatmullRom);

            let mut buf = Vec::new();
            let mut cursor = Cursor::new(&mut buf);
            let encoder = JpegEncoder::new_with_quality(&mut cursor, 30);
            if let Err(e) = resized.write_with_encoder(encoder) {
                self.debug_print(&format!("縮小時JPEGエンコードエラー: {}", e));
                return Ok(image_data.to_vec());
            }
            result_data = buf;
            self.debug_print(&format!(
                "追加縮小 {:.1}: {}x{}, {} bytes",
                scale_factor,
                new_width,
                new_height,
                result_data.len()
            ));

            if result_data.len() <= limits.max_file_size {
                self.debug_print(&format!(
                    "完了: {} bytes (スケール: {:.1})",
                    result_data.len(),
                    scale_factor
                ));
                return Ok(result_data);
            }

            scale_factor -= 0.1;
        }

        self.debug_print(&format!("最終: {} bytes", result_data.len()));
        Ok(result_data)
    }

    #[allow(dead_code)]
    pub fn resize_image_file(&self, file_path: &str, sns_type: &str) -> anyhow::Result<Vec<u8>> {
        let original_data = std::fs::read(file_path)?;
        self.resize_image_data(&original_data, sns_type)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use image::{ImageBuffer, Rgb};

    fn create_dummy_image(width: u32, height: u32) -> Vec<u8> {
        let img = ImageBuffer::from_fn(width, height, |_, _| Rgb([255, 0, 0]));
        let dynamic_img = DynamicImage::ImageRgb8(img);
        let mut buf = Vec::new();
        let mut cursor = std::io::Cursor::new(&mut buf);
        dynamic_img
            .write_to(&mut cursor, image::ImageFormat::Jpeg)
            .unwrap();
        buf
    }

    #[test]
    fn test_resize_no_need() {
        let resizer = ImageResizer::new(false);
        let data = create_dummy_image(100, 100);
        let result = resizer.resize_image_data(&data, "bluesky").unwrap();
        assert_eq!(result.len(), data.len());
    }

    #[test]
    fn test_resize_large_image() {
        let resizer = ImageResizer::new(false);
        let data = create_dummy_image(3000, 3000);
        let result = resizer.resize_image_data(&data, "bluesky").unwrap();

        let img = image::load_from_memory(&result).unwrap();
        assert!(img.width() <= 2000);
        assert!(img.height() <= 2000);
        assert!(result.len() <= 1024 * 1024);
    }

    /// JPEG/PNG以外(ここではBMP)は上限内でもJPEGへ変換されることを確認する。
    /// upload_mimeがimage/jpegとして送るため、実体もJPEGである必要がある。
    #[test]
    fn test_non_jpeg_png_converted_to_jpeg() {
        let resizer = ImageResizer::new(false);
        // 100x100の小さなBMP(上限内)を作る
        let img = ImageBuffer::from_fn(100, 100, |_, _| Rgb([0, 128, 255]));
        let dynamic_img = DynamicImage::ImageRgb8(img);
        let mut data = Vec::new();
        let mut cursor = std::io::Cursor::new(&mut data);
        dynamic_img
            .write_to(&mut cursor, image::ImageFormat::Bmp)
            .unwrap();
        assert_eq!(image::guess_format(&data).unwrap(), image::ImageFormat::Bmp);

        let result = resizer.resize_image_data(&data, "x").unwrap();
        // 出力はJPEGになっている
        assert_eq!(
            image::guess_format(&result).unwrap(),
            image::ImageFormat::Jpeg
        );
    }

    /// デコードできない(対応外形式の)データはエラーになり、
    /// 元バイトをそのまま返さないことを確認する。
    #[test]
    fn test_undecodable_returns_err() {
        let resizer = ImageResizer::new(false);
        let garbage = vec![0u8, 1, 2, 3, 4, 5, 6, 7];
        let result = resizer.resize_image_data(&garbage, "x");
        assert!(result.is_err());
    }
}
